euphonic/fast_adaptive_broadening.py

```diff
@@ -0,0 +1,162 @@
+        sigma_factors = mode_widths[masked_block]/mode_width_samples[i-1]
+        lower_mix = np.polyval(lower_coeffs, sigma_factors)
+        upper_mix = 1-lower_mix
+
+        n_spec = np.zeros(len(mode_widths))
+        n_spec[masked_block] = freqs[masked_block]
+        low_weights = np.zeros(len(mode_widths))
+        up_weights = np.zeros(len(mode_widths))
+        low_weights[masked_block] = lower_mix
+        up_weights[masked_block] = upper_mix
+
+        lower_hist, _ = np.histogram(
+            n_spec, bins=dos_bins, weights=low_weights*n_weights*mode_weights/bin_width)
+        upper_hist, _ = np.histogram(
+            n_spec, bins=dos_bins, weights=up_weights*n_weights*mode_weights/bin_width)
+
+        scaled_data_matrix[:, i-1] += lower_hist
+        scaled_data_matrix[:, i] += upper_hist
+
+    dos = np.sum([convolve(scaled_data_matrix[:, i], kernels[i],
+                 mode="same") for i in range(0, n_kernels)], 0)
+    return dos
+
+
+def gaussian(xvals: np.ndarray,
+             sigma: np.ndarray,
+             centre: Optional[int] = 0) -> np.ndarray:
+    """
+    Evaluates the Gaussian function.
+
+    Parameters
+    ----------
+    xvals
+        Float ndarray. Points at which the Gaussian function should be
+        evaluated
+    sigma
+        Float ndarray. Specifies the standard deviation for the gaussian
+    centre
+        Optional integer value that sets the centre of the gaussian
+
+    Returns
+    -------
+    gauss_eval
+        Float ndarray containing the values of the evaluated gaussian function
+    """
+    # evaluate gaussian function with defined sigma and center at x
+    gauss_eval = np.exp(-0.5 * ((xvals - centre) / sigma)**2) \
+                    / (sigma * np.sqrt(2 * np.pi))
+    return gauss_eval
+
+
+def find_coeffs(spacing: float) -> np.ndarray:
+    """"
+    Function that, for a given spacing value, gives the coefficients of the
+    polynomial which decsribes the relationship between sigma and the
+    linear combination weights determined by optimised interpolation
+
+    Parameters
+    ----------
+    spacing
+        Scalar float. The spacing value between sigma samples at which
+        the gaussian kernel is exactly calculated.
+
+    Returns
+    -------
+    coeffs
+        Array containing the polynomial coefficients, with the highest
+        power first
+    """
+    sigma_values = np.linspace(1, spacing, 10)
+    x_range = np.linspace(-10, 50, 101)
+
+    def gaussian_mix(xvals, weight):
+        # Return a linear combination of two Gaussians with weights
+        return (weight * gaussian(xvals, sigma=1)
+                + (1-weight) * gaussian(xvals, sigma=spacing))
+
+    lower_mix = np.zeros(len(sigma_values))
+
+    for i, s_val in enumerate(sigma_values):
+        actual_gaussian = gaussian(x_range, s_val)
+        mixl, _ = curve_fit(gaussian_mix, x_range,
+                            ydata=actual_gaussian, p0=[0.5], bounds=(0, 1))
+        lower_mix[i] = mixl[0]
```

Comment:
It looks like if we need it there is a speedup to had by switching from nonlinear to linear-least-squares fitting. (scipy.linalg.lstsq)

```

In [128]: %timeit lstsq(np.array([norm.pdf(x_range, scale=1), norm.pdf(x_range, scale=1.5)]).T, actual_
     ...: gaussian)
242 µs ± 1.88 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [129]: %timeit mixl, _ = curve_fit(gaussian_mix, x_range, ydata=actual_gaussian, p0=[0.5], bounds=(0
     ...: , 1))
2.96 ms ± 22.1 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

The catch is that we don't get to enforce bounds any more, so might have to watch out for strange negative solutions.

Reply:
See this [notebook](https://github.com/pace-neutrons/Broadening-Methods/blob/main/Compare%20least%20squares%20solvers.ipynb) for a comparison of different least-squares solvers - the fastest choice appears to be `scipy.optimize.nnls`  which is a non-negative least squares solver and so this was used to replace `curve_fit`